### ISSUE TITLE
Blazor version 8.0.5 - If I have even one connected client, shutdown of the app is 30 seconds

### ISSUE THREAD

**Issue by alkampfergit** [dotnet/aspnetcore#58947]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

This is probably a duplicate of https://github.com/dotnet/aspnetcore/issues/53165 . Our application, whenever there  is at least one client connected needs 30 seconds from when the Asp.NET core receve the IsClosing from when the app is closed.

This is really annoying because we have software running as windows service, and having this long timeout sometimes generates problem.

We took the timing with IHostedLifecycleService and whenever there is a client connected the time between closing and closed is always 30 seconds (with a difference of max 100 milliseconds). 

We are using version 8.0.5, and issue [#53165](https://github.com/dotnet/aspnetcore/issues/53165) is closed so I'm expecting not to have this problem. 

### Expected Behavior

When I shutdown my asp.net core application I do not want signalr block the shutdown for 30 seconds, because it causes problem.

I'm expecting also the possibility to force signalr to close everything and letting the application shutdown correctly.

### Steps To Reproduce

I do not have a minimal reproduction, it happens always on our software.

### .NET Version

8.0.404

**Comment by BrennanConroy** [comment]
Show server logs please, similar to what is in the issue you linked to
https://github.com/dotnet/aspnetcore/issues/53165#issuecomment-1884978632

**Comment by alkampfergit** [comment]
Ok here is what happened, this is the output of the console

[10:11:42 INF][Jarvis.Host.Support.WebAppLifecycleManagerHostedService] ################### StopAsync completed in 7180 ms.
[10:11:42 WRN][Orleans.Messaging] Dropping message Request [S10.0.0.39:40101:90666655 sys.svc.user.95B6CAE6/10.0.0.39:40101@90666655]->[S10.0.0.39:40101:90666655 management/0] Orleans.Runtime.IManagementGrain.GetRuntimeStatistics(Orleans.Runtime.SiloAddress[]) #5815 since this silo is blocking application messages
[10:12:04 INF][Microsoft.AspNetCore.Routing.EndpointMiddleware] Executed endpoint '/notification'
[10:12:04 INF][Microsoft.AspNetCore.Routing.EndpointMiddleware] Executed endpoint '/chat'
[10:12:04 INF][Microsoft.AspNetCore.Hosting.Diagnostics] Request finished HTTP/2 CONNECT https://jarvis-uniquehost.codewrecks.com/notification?id=ewhvNu2JeTu1Jf3oaH3qRg - 200 null null 29973.9655ms
[10:12:04 INF][Microsoft.AspNetCore.Hosting.Diagnostics] Request finished HTTP/2 CONNECT https://jarvis-uniquehost.codewrecks.com/chat?id=ARw59xut9nTlSsFXjp_orw - 200 null null 29974.0106ms
[10:12:04 INF][Microsoft.AspNetCore.Server.Kestrel] Connection id "0HN85ALL3T7BF", Request id "0HN85ALL3T7BF:00000035": the application completed without reading the entire request body.
[10:12:04 INF][Microsoft.AspNetCore.Server.Kestrel] Connection id "0HN85ALL3T7BF", Request id "0HN85ALL3T7BF:00000037": the application completed without reading the entire request body.
[10:12:04 INF][Jarvis.Host.Support.WebAppLifecycleManagerHostedService] ###################  StoppedAsync completed in 30015 ms.
[10:12:04 INF][Jarvis.Host.Support.WebAppLifecycleManagerHostedService] ################### OnStopped called in 30017 ms.

The first line shows that we closed all of our components after 7180 ms from when the IHost received CTRL+C (mainly is orleans), then I have no log up to 30017 ms when all other logs arrives. As you can see we have logs of our two endpoints for signalr /notification and /chat

I've uploaded trace file where I trace evertying that happened in signalr and asp.net

[trace_extended20241115.log](https://github.com/user-attachments/files/17772540/trace_extended20241115.log)

**Comment by BrennanConroy** [comment]
I see WebAppLifecycleManagerHostedService in your logs, is this an `IHost`/`IWebHost` implementation and does it implement `IHostApplicationLifetime`? SignalR uses `IHostApplicationLifetime.ApplicationStopping` to trigger connection disconnects.

If that doesn't help figure out the problem, could you enable trace logs for kestrel, hosting, signalr, and http.connections?

**Comment by alkampfergit** [comment]
The WebAppLifecycleManagerHostedService  is a simple IHostedLifecycleService that is used to detect application closing and closing all of the services of our application gracefully. Internally it uses IHostApplicationLifetime to register The application stopping, and it is in that specific method that we send disconnect explicitly to signalr client. Without explicit shutdown, the application closes always in 30 seconds.

**Comment by BrennanConroy** [comment]
> could you enable trace logs for kestrel, hosting, signalr, and http.connections?

Were you able to capture these logs?

**Comment by alkampfergit** [comment]
Attacched is another file log, let me know if you need still more loggers.

[trace_extended20241212.log](https://github.com/user-attachments/files/18108440/trace_extended20241212.log)

**Comment by alkampfergit** [comment]
Yes you are true, we heavily depend on signalr so clients after a
disconnect, immediately try to reconnect.

We solved in a simple way, when we want to close, we send a message to all
clients to ask to disconnect , this seems to fix the issue

Il gio 12 dic 2024, 18:53 Brennan ***@***.***> ha scritto:

**Comment by noelex** [comment]
I'm also facing this issue after upgrading to .NET 9.0, but with gRPC server streaming calls. The shutdown of the server application will be blocked if there's any active server streaming call when the shutdown is initiated.

Even if I register to the `IHostApplicationLifetime.ApplicationStopping` signal in the streaming method and abort the method when shutdown is requested, the shutdown will still be blocked for about 30 seconds.

My workaround for this is to block the shutdown until all ongoing requests are finished, so that ASP.NET Core can start shutdown without being blocked by the streaming call:

```csharp
internal interface IShutdownSignal
{
    void AddBlocker();
    void RemoveBlocker();
    CancellationToken ApplicationStopping { get; }
}

sealed class ShutdownSignal : IShutdownSignal, IDisposable
{
    private readonly CancellationTokenSource _cts = new();
    private readonly SemaphoreSlim _sem = new(1, 1);

    private long _blockers;

    public ShutdownSignal(IHostApplicationLifetime lifetime)
    {
        _logger = logger;
        lifetime.ApplicationStopping.Register(static (state) =>
        {
            var self = (ShutdownSignal)state!;
            self._cts.Cancel();
            self._sem.Wait();
        }, this);
    }

    public CancellationToken ApplicationStopping => _cts.Token;

    public void AddBlocker()
    {
        if (Interlocked.Increment(ref _blockers) == 1)
        {
            _sem.Wait();
        }
    }

    public void RemoveBlocker()
    {
        if (Interlocked.Decrement(ref _blockers) == 0)
        {
            _sem.Release();
        }
    }

    public void Dispose()
    {
        _cts.Dispose();
        _sem.Dispose();
    }
}
```

```csharp
builder.Services.AddSingleton<IShutdownSignal, ShutdownSignal>();

await using var app = builder.Build();

var ss = app.Services.GetRequiredService<IShutdownSignal>();
app.Use(async (context, next) =>
{
    try
    {
        ss.AddBlocker();
        await next(context);
    }
    finally
    {
        ss.RemoveBlocker();
    }
});
```

```csharp
private readonly IShutdownSignal _signal;

public override async Task ServerStream(RequestMessage request, IServerStreamWriter<ResponseMessage> responseStream, ServerCallContext context)
{
    using var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken, _signal.ApplicationStopping);
    await ServerStreamImpl(request, response, cts.Token);
}
```

**Comment by KnapSac** [comment]
@BrennanConroy you mentioned that a fix was planned, any updates on when that fix might land? We can work around it by sending a shutdown message, so we're not blocked currently, but it does feel like a bit of a hacky solution.

**Comment by beppemarazzi** [comment]
While waiting for an official solution i'm doing this in my blazor server projects (comments are welcome):

```
internal class FixSignalRReconnectionPreventingShutdownMiddleware(
    RequestDelegate next,
    IHostApplicationLifetime lifetime,
    ILogger<FixSignalRReconnectionPreventingShutdownMiddleware> logger
    )
{
    //see https://github.com/dotnet/aspnetcore/issues/58947#issuecomment-2539625627
    public Task InvokeAsync(HttpContext context)
    {
        if (lifetime.ApplicationStopping.IsCancellationRequested)
        {
            if (context.Request.Path.StartsWithSegments("/_blazor/negotiate"))
            {
                logger.LogDebug("Rejecting SignalR request during shutdown to avoid shutdown delay (https://github.com/dotnet/aspnetcore/issues/58947#issuecomment-2539625627).");
                //option1 : error at http level: return 503
                //context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
                context.Response.WriteAsJsonAsync(new
                {
                    //option2 : redirect the SignalR negotiate request to a dummy url to avoid persistent retries
                    url = "http://localhost/server_is_shutting_down",

                    //option3 : error at SignalR negotiation level
                    //error = "Server is shutting down."
                });
                return Task.CompletedTask;
            }
        }
        return next(context);
    }
}
```

**Comment by juho-hanhimaki** [comment]
Hoping this gets fixed in .NET 11. Had some good head scratching time trying to figure out what's holding my app hostage for 30 seconds.
